pkg/runtime/grpc/k8s-portfwd-dialer.go

```diff
@@ -0,0 +1,157 @@
+// Copyright 2023 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package grpcruntime
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"path"
+	"strconv"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/httpstream"
+	"k8s.io/client-go/tools/portforward"
+	"k8s.io/client-go/transport/spdy"
+
+	"github.com/inspektor-gadget/inspektor-gadget/cmd/kubectl-gadget/utils"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/factory"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-service/api"
+)
+
+type k8sPortFwdDialer struct {
+	io.Writer
+	io.Reader
+	conn       httpstream.Connection
+	stream     httpstream.Stream
+	podName    string
+	cancelFunc func()
+}
+
+// NewK8SPortFwdConn connects to a Pod using PortForwarding via the Kubernetes API Server
+func NewK8SPortFwdConn(ctx context.Context, pod target, timeout time.Duration) (net.Conn, error) {
+	conn := &k8sPortFwdDialer{}
+
+	config, err := utils.KubernetesConfigFlags.ToRESTConfig()
+	if err != nil {
+		return nil, fmt.Errorf("creating RESTConfig: %w", err)
+	}
+
+	// set GroupVersion and NegotiatedSerializer for RESTClient
+	factory.SetKubernetesDefaults(config)
+
+	conn.podName = pod.addressOrPod
+
+	config.Timeout = timeout
+
+	transport, upgrader, err := spdy.RoundTripperFor(config)
+	if err != nil {
+		return nil, fmt.Errorf("creating roundtripper: %w", err)
+	}
+
+	targetURL, err := url.Parse(config.Host)
+	if err != nil {
+		return nil, fmt.Errorf("parsing restConfig.Host: %w", err)
+	}
+
+	targetURL.Path = path.Join(
+		"api", "v1",
+		"namespaces", "gadget",
+		"pods", conn.podName,
+		"portforward",
+	)
```

Comment:
I'm wondering that's the advantage over a simple `fmt.Sprintf("/api/v1/....%s.....", conn.podName)`

Reply:
I don't really see one in this case - I believe the code handling the port forwarding in k8s used this, so I went with it.

Reply:
> `Join()` is platform agnostic, in case we run this code on Windows node (I do not know if it is possible and am wondering about knowing).

That'd be `filepath`, then. Here we're building a URL with the `path` package, so it's only handling forward slashes.